File: spec/java/src/io/kaitai/struct/spec/TestEofExceptionBitsBe.java
package io.kaitai.struct.spec;

import io.kaitai.struct.testformats.EofExceptionBitsBe;
import org.testng.annotations.Test;
import static org.testng.Assert.*;
public class TestEofExceptionBitsBe extends CommonSpec {
    @Test
    public void testEofExceptionBitsBe() throws Exception {
        EofExceptionBitsBe r = EofExceptionBitsBe.fromFile(SRC_DIR + "nav_parent_switch.bin");
        assertThrowsEofError(new ThrowingRunnable() {
            @Override
            public void run() throws Throwable {
                r._read();
            }
        });
        assertIntEquals(r._io().pos(), 1);
        assertIntEquals(r.preBits(), 0);
        assertTrue(r._attrStart.containsKey("failBits"));
        assertIntEquals(r._attrStart.get("failBits"), 1);
        assertIntEquals(r.failBits(), 0); // actually uninitialized, hence 0 in Java
        assertFalse(r._attrEnd.containsKey("failBits"));
    }
}

File: spec/python/spec/test_eof_exception_bits_be.py
import unittest
from testformats.eof_exception_bits_be import EofExceptionBitsBe
import kaitaistruct

class TestEofExceptionBitsBe(unittest.TestCase):
    def test_eof_exception_bits_be(self):
        with EofExceptionBitsBe.from_file('src/nav_parent_switch.bin') as r:
            with self.assertRaises(EOFError) as cm:
                r._read()
            self.assertEqual(str(cm.exception), u"requested 3 bytes, but only 2 bytes available")

            self.assertEqual(r._io.pos(), 1)

            self.assertEqual(r.pre_bits, 0b0000000)  # 0b0000_000
            self.assertIn('start', r._debug['fail_bits'])
            self.assertEqual(r._debug['fail_bits']['start'], 1)
            self.assertFalse(hasattr(r, 'fail_bits'))
            self.assertNotIn('end', r._debug['fail_bits'])
